lzip: add mingw support

seanybaggins opened the pull request:
Part of a greater effort to get qtbase and qtmultimedia cross compiling for windows.
- NixOS#274274
- NixOS#272538

## Description of changes
Added mingw support for lzip
<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

seanybaggins commented:
@ofborg build pkgs.pkgsCross.mingw32.lzip

seanybaggins commented:
@ofborg build pkgs.pkgsCross.mingwW64.lzip

seanybaggins commented:
@wegank @Artturin @RaitoBezarius

wegank commented on pkgs/tools/compression/lzip/default.nix:
```diff
@@ -27,6 +30,11 @@ stdenv.mkDerivation rec {
   doCheck = true;
   enableParallelBuilding = true;
 
+  installPhase = lib.optionalString stdenv.hostPlatform.isMinGW ''
```
This will break `lzip` on other platforms, since they now have `installPhase = ""`, right?

seanybaggins replied:
Oooof. It looks that way. What I wanted was in its not a mingw build, use the default installPhase of builder. Otherwise use the installPhase that I have specified.

I thought if installPhase was empty it would use the default, looking at the rebuild count, that assumption clearly was not true.

wegank replied:
It would indeed be empty if `installPhase` evaluates to `null`.

seanybaggins replied:
I think I started by trying to patch the Makefile.in file but ran into issues. Could be mistaken though, its been a little bit. I think what I would like to try and do is fix my logic at the installPhase

seanybaggins replied:
Fixed.

seanybaggins commented:
@ofborg build pkgs.pkgsCross.mingw32.lzip

seanybaggins commented:
@ofborg build pkgs.pkgsCross.mingwW64.lzip

wegank commented on pkgs/tools/compression/lzip/default.nix:
```diff
@@ -27,6 +30,13 @@ stdenv.mkDerivation rec {
 
+  installPhase =
+    if stdenv.hostPlatform.isMinGW then ''
+      sed -i '/^\s*$(INSTALL_PROGRAM) .\/$(progname) "\$(DESTDIR)$(bindir)\/$(progname)"$/s/$(progname)/$(progname).exe/g' Makefile
```
I'd prefer this line to go to `postPatch`, and see if `installPhase` can be dropped completely and `outputs` made unconditional.

seanybaggins replied:
Okay managed to get it the way you suggested. Way cleaner.

seanybaggins committed:
lzip: add mingw support
Part of a greater effort to get qtbase and qtmultimedia cross compiling
for windows.

seanybaggins commented:
@ofborg build pkgs.pkgsCross.mingw32.lzip

seanybaggins commented:
@ofborg build  pkgs.pkgsCross.mingwW64.lzip

wegank reviewed (approved):
LGTM. There is currently [a patch on the mailing list](https://lists.nongnu.org/archive/html/lzip-bug/2023-08/msg00005.html), which is stale and is not trivial anyway, because the patch there has to be unconditional, i.e. extension must be detected automatically.

Still, a possible improvement could be to apply `substituteAll` on the patch with `stdenv.hostPlatform.extensions.executable`. I'll leave it to the other reviewers to decide if they wish.

Artturin commented on pkgs/tools/compression/lzip/mingw-install-exe-file.patch:
```diff
@@ -0,0 +1,13 @@
+ install-bin : all
+ 	if [ ! -d "$(DESTDIR)$(bindir)" ] ; then $(INSTALL_DIR) "$(DESTDIR)$(bindir)" ; fi
+-	$(INSTALL_PROGRAM) ./$(progname) "$(DESTDIR)$(bindir)/$(progname)"
++	$(INSTALL_PROGRAM) ./$(progname).exe "$(DESTDIR)$(bindir)/$(progname).exe"
```
```suggestion
+	$(INSTALL_PROGRAM) ./$(progname)$(EXEEXT) "$(DESTDIR)$(bindir)/$(progname)$(EXEEXT)"
```

https://www.gnu.org/software/automake/manual/html_node/EXEEXT.html

wegank replied:
> However, Automake cannot apply this rewriting to configure substitutions

There's no configure.ac, so the change probably doesn't work.

Artturin replied:
Ok thanks for trying.
